Prevent rollbacks on failing registration

laenion opened the pull request:
There are currently two open bugs for CaaSP which cause a rollback of a system to the previous release after successful migration (https://bugzilla.suse.com/show_bug.cgi?id=1108618 & https://bugzilla.suse.com/show_bug.cgi?id=1113048). This will happen if the SUSEConnect call was not successful, i.e. the release information is still the old one.

These two commits try to make the behaviour more reliable to prevent these situations.

thkukuk commented on systemd/rollback.service:
```diff
@@ -1,8 +1,8 @@
 DefaultDependencies=no
-Requires=network-online.target
-After=local-fs.target network-online.target
+After=local-fs.target systemd-networkd-wait-online.service network.service
```
We don't use systemd-networkd and network.service does not mean the network is online, so how do you make sure that we only run the service if the network is online?

laenion commented:
The services were taken from the cloud-init service file (the online part). While not necessary the networkd part was copied to match the criteria.
As for the network.service part: Yes, it isn't guaranteed that the network is online, but neither it is on the network-online target - that's the problem of the first bug report. To be able to run before the online cloud-init part this dependency has to be loosened, as there would be cyclic dependencies otherwise (as cloud-init itself tries to run *before* the online target).

thkukuk commented:
On Tue, Oct 30, Ignaz Forster wrote:

> The services were taken from the cloud-init service file (the online part).
> While not necessary the networkd part was copied to match the criteria.
> As for the network.service part: Yes, it isn't guaranteed that the network is
> online, but neither it is on the network-online target - that's the problem of
> the first bug report. To be able to run before the online cloud-init part this
> dependency has to be loosened, as there would be cyclic dependencies otherwise
> (as cloud-init itself tries to run before the online target).

This will not work. rollback.service requires network access. If it is
running before cloud-init.service, and cloud-init.service must be run
before network is online, rollback.service will never have access to
the network and will never work except by accident.
The cyclic dependency systemd founds is correct, your workaround will
not work reliable and make things only worse.

-- 
Thorsten Kukuk, Distinguished Engineer, Senior Architect SLES & MicroOS
SUSE LINUX GmbH, Maxfeldstr. 5, 90409 Nuernberg, Germany
GF: Felix Imendoerffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nuernberg)

laenion commented:
> This will not work. rollback.service requires network access. If it is running before cloud-init.service, and cloud-init.service must be run before network is online, rollback.service will never have access to the network and will never work except by accident. The cyclic dependency systemd founds is correct, your workaround will not work reliable and make things only worse.

Agreed. According to https://cloudinit.readthedocs.io/en/latest/topics/boot.html#boot-stages the _cloud-init.service_ matches the **Network** stage. According to _/etc/cloud/cloud.cfg_ zypper configuration should run in the later **Config** stage (represented by _cloud-config.service_, which will run after _network-online.target_). Sorry for not reading the documentation beforehand - I'll change the dependencies accordingly.

laenion committed:
Retry network connection if it doesn't work yet

Wicked may return even if the network was not completely set up yet (see e.g.
bsc#1108618). Unforutnately systemd also doesn't support restarting oneshot
services yet, so implement a manual retry (and return with a failure status
if SUSEConnect wasn't successful).

laenion committed:
Run before any other services calling zypper

After a migration zypper must not be called before the product was registered
correctly, otherwise bad things will happen, including downgrades of the
complete product (see bsc#1113048).
Both cloud-init and salt may trigger reloading zypper repositories, so make
sure to run before them.

laenion commented:
Changed the commit according to the previous comment.

The new startup sequence for reference:
![new_startup](https://user-images.githubusercontent.com/2150415/48206828-8cecbc00-e36f-11e8-9af4-bbbb239bb733.png)

laenion commented:
@jsrain: Since Thorsten is on vacation and you are the only other project maintainer: Would it be possible to review these changes?
